spec/Section 3 -- Type System.md

```diff
@@ -1340,14 +1357,281 @@ Union type extensions have the potential to be invalid if incorrectly defined.
 
 1. The named type must already be defined and must be a Union type.
 2. The member types of a Union type extension must all be Object base types;
-   Scalar, Interface and Union types must not be member types of a Union.
-   Similarly, wrapping types must not be member types of a Union.
+   Scalar, Interface, Union and Tagged types must not be member types of a
+   Union.  Similarly, wrapping types must not be member types of a Union.
 3. All member types of a Union type extension must be unique.
 4. All member types of a Union type extension must not already be a member of
    the original Union type.
 5. Any non-repeatable directives provided must not already apply to the
    original Union type.
 
+
+## Tagged Types
+
+TaggedTypeDefinition : Description? tagged Name Directives[Const]? TaggedMembersDefinition?
+
+TaggedTypeExtension :
+  - extend tagged Name Directives[Const]? TaggedMembersDefinition
+  - extend tagged Name Directives[Const]
+
+TaggedMembersDefinition : { TaggedMemberDefinition+ }
+
+TaggedMemberDefinition : Description? Name : Type Directives[Const]?
+
+Tagged types represent a list of possible named members, exactly one of which
+must be present. This resulting member must yield a value of a specific type.
+In outputs, like for Objects, Tagged types should be serialized as ordered
+maps: the queried member names (or aliases) that are present are the keys and
+the result of evaluating the member is the value, ordered by the order in which
+they appear in the query. Note that Tagged types may be objects with more than
+one key even though they guarantee that exactly one member is present, this is
+because aliases may be used, and `__typename` may also be queried (and will
+return the name of the tagged type).
+
+All members defined within a Tagged type must not have a name which begins with
+{"__"} (two underscores), as this is used exclusively by GraphQL's
+introspection system.
+
+For example, a type `StringFilter` could be described as:
+
+```graphql example
+tagged StringFilter {
+  startsWith: String!
+  contains: String!
+  lengthAtLeast: Int!
+}
+```
+
+In this case we're representing exactly one of the following:
+
+- an object containing a single key `startsWith` with an {String} value
+- an object containing a single key `contains` with a {String} value
+- an object containing a single key `lengthAtLeast` with an {Int} value
+
+The `StringFilter` Tagged type is valid as both an input and output type, but
+this is not true of all tagged types. If a Tagged type has a member whose type
+is only valid for output, then the Tagged type is only valid for output. If a
+Tagged type has a member whose type is only valid for input, then the Tagged
+type is only valid for input. If a Tagged type has a member that is only valid
+for input, and another member that's only valid for output, then it is not a
+valid Tagged type.
+
+A member of a Tagged type may be a Scalar, Enum, Input Object type, Object
```

Comment:
Is this list just literally any type? If not it may be clearer to list what is excluded, not what is included.

Reply:
It's all of the currently defined types, yes, but I worried that stating "any type" might limit future option value. You're right though, clarity wins. Fixed :+1: